linux/ibus-keyman/src/engine.c

```diff
@@ -388,8 +388,8 @@ ibus_keyman_engine_constructor (GType                   type,
         g_warning("problem creating km_kbp_state");
     }
     for (int i =0; i < 3; i++) {
-        km_kbp_cp_dispose(keyboard_opts[i].key);
-        km_kbp_cp_dispose(keyboard_opts[i].value);
+        g_free((km_kbp_cp *)keyboard_opts[i].key);
```

Comment:
Did you allocate these or do they come from KPAPI? If they come from KPAPI then can't use g_free (check with @tim-eves )

Reply:
It looks like he allocated those through `g_utf8_to_utf16()` between line 327-344. The original code would have technically been wrong, since `km_kbp_*_dispose` calls are only supposed to be used on pointers from the API not thing allocated by other functions.
BTW @glasseyes you could make keyboard_opts array a regular stack allocated C array, provided the last entry is `{0,0,0}` or `KM_KBP_OPTIONS_END`.

Reply:
I had it stack allocated before but it was simpler initialising it with gnew0 and not having to worry about creating the last entry.